# Домашнее задание 2 (5 баллов).

Все задания ниже имеют равный вес (5/16).

### Мосалев Максим Сергеевич ОМО-ИАД-7

In [258]:
import pandas as pd
import numpy as np

In [277]:
#чтобы скачать и разархвировать файл с данными:
import urllib.request
import zipfile
destination = 'Data.zip'
url = 'https://github.com/hse-ds/iad-intro-ds/raw/master/2023/homeworks/Data.zip'
urllib.request.urlretrieve(url, destination)

with zipfile.ZipFile('Data.zip', 'r') as zip_ref:
    zip_ref.extractall()

#### Описание данных

В папке Dat (https://github.com/hse-ds/iad-intro-ds/blob/master/2023/homeworks/Data.zip) находится информация о студентах. Всего 10 групп студентов. Файлы делятся на две категории:
    * Students_info_i - информация о студентах из группы i
    * Students_marks_i - оценки студентов из группы i за экзамены

### Одно из важных достоинств pandas $-$ это удобные методы реляционного взаимодействия с данными, аналогичные, например, возможностям SQL для слияния и конкатенации таблиц: merge, join, concat. Наличие готовых методов позволяет не реализовывать самостоятельно поэлементную обработку данных и оперировать сразу целыми таблицами данных.

Подробно об этих методах посмотрите тут: https://www.kaggle.com/residentmario/renaming-and-combining#Combining

#### Задание 1. Соберите всю информацию о студентах в одну таблицу df. В получившейся таблице должна быть информация и оценки всех студентов из всех групп. Напечатайте несколько строк таблицы для демонстрации результата.¶

In [260]:
df = pd.DataFrame()
for i in range(0, 10):
    df_inf = pd.read_csv('Data\Students_info_' + str(i) + '.csv')
    df_mrk = pd.read_csv('Data\Students_marks_' + str(i) + '.csv')
    df_i = pd.merge(df_inf, df_mrk, on = 'index')
    df = pd.concat([df, df_i])
df.head(10)

index  gender race/ethnicity parental level of education         lunch  \
0      0  female        group B           bachelor's degree      standard   
1      1  female        group C                some college      standard   
2      2  female        group B             master's degree      standard   
3      3    male        group A          associate's degree  free/reduced   
4      4    male        group C                some college      standard   
5      5  female        group B          associate's degree      standard   
6      6  female        group B                some college      standard   
7      7    male        group B                some college  free/reduced   
8      8    male        group D                 high school  free/reduced   
9      9  female        group B                 high school  free/reduced   

  test preparation course   group  math score  reading score  writing score  
0                    none  group1          72             72             74  
1               completed  group1          69             90             88  
2                    none  group1          90             95             93  
3                    none  group1          47             57             44  
4                    none  group1          76             78             75  
5                    none  group1          71             83             78  
6               completed  group1          88             95             92  
7                    none  group1          40             43             39  
8               completed  group1          64             64             67  
9                    none  group1          38             60             50

#### Задание 2. Удалите столбец index у полученной таблицы. Напечатайте первые 10 строк таблицы.

In [261]:
df = df.drop(columns = 'index')
df.head(10)

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   
5  female        group B          associate's degree      standard   
6  female        group B                some college      standard   
7    male        group B                some college  free/reduced   
8    male        group D                 high school  free/reduced   
9  female        group B                 high school  free/reduced   

  test preparation course   group  math score  reading score  writing score  
0                    none  group1          72             72             74  
1               completed  group1          69             90             88  
2                    none  group1          90             95             93  
3                    none  group1          47             57             44  
4                    none  group1          76             78             75  
5                    none  group1          71             83             78  
6               completed  group1          88             95             92  
7                    none  group1          40             43             39  
8               completed  group1          64             64             67  
9                    none  group1          38             60             50

#### Задание 3. Выведите на экран размеры полученной таблицы

In [262]:
df.shape

(1000, 9)

#### Задание 4. Выведите на экран статистические характеристики числовых столбцов таблицы (минимум, максимум, среднее значение, стандартное отклонение)

In [263]:
df.describe().loc[['min', 'max', 'mean', 'std']]

math score  reading score  writing score
min      0.00000      17.000000      10.000000
max    100.00000     100.000000     100.000000
mean    66.08900      69.169000      68.054000
std     15.16308      14.600192      15.195657

#### Задание 5. Проверьте, есть ли в таблице пропущенные значения

In [264]:
df.isna().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
group                          0
math score                     0
reading score                  0
writing score                  0
dtype: int64

Значит, в таблице нет пропущенных значение 

#### Задание 6. Выведите на экран средние баллы студентов по каждому предмету (math, reading, writing)

In [265]:
df.describe().loc['mean']

#или
#print('Math average score = ', df['math score'].mean())
#print('Reading average score = ', df['reading score'].mean())
#print('Writing average score = ', df['writing score'].mean())

math score       66.089
reading score    69.169
writing score    68.054
Name: mean, dtype: float64

**Задание 7. Как зависят оценки от того, проходил ли студент курс для подготовки к сдаче экзамена (test preparation course)? Выведите на экран для каждого предмета в отдельности средний балл студентов, проходивших курс для подготовки к экзамену и не проходивших курс.**

In [266]:
df[['test preparation course', 'math score', 'reading score', 'writing score']].groupby(by='test preparation course').mean()

math score  reading score  writing score
test preparation course                                          
completed                 69.695531      73.893855      74.418994
none                      64.077882      66.534268      64.504673

У тех студентов, которые не проходили курс для подготовки, средний балл по всем предметам ниже, чем у тех студентов, которые проходили этот курс. 

**Задание 8. Выведите на экран все различные значения из столбца lunch.**

In [267]:
df.lunch.unique()

array(['standard', 'free/reduced'], dtype=object)

**Задание 9. Переименуйте колонку "parental level of education" в "education", а "test preparation course" в "test preparation" с помощью метода pandas rename**
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rename.html

In [268]:
df = df.rename(columns = {"parental level of education": "education", "test preparation course": "test preparation"})
df.head(5)

gender race/ethnicity           education         lunch test preparation  \
0  female        group B   bachelor's degree      standard             none   
1  female        group C        some college      standard        completed   
2  female        group B     master's degree      standard             none   
3    male        group A  associate's degree  free/reduced             none   
4    male        group C        some college      standard             none   

    group  math score  reading score  writing score  
0  group1          72             72             74  
1  group1          69             90             88  
2  group1          90             95             93  
3  group1          47             57             44  
4  group1          76             78             75

**Зафиксируем минимальный балл для сдачи экзамена**

In [269]:
passmark = 50

**Задание 10. Ответьте на вопросы:**
    * Какая доля студентов сдала экзамен по математике (passmark >= 50)?
    * Какая доля студентов, проходивших курс подготовки к экзамену, сдала экзамен по математике?
    * Какая доля женщин, не проходивших курс подготовки к экзамену, не сдала экзамен по математике? 

In [270]:
print(df[df['math score'] >= passmark].shape[0] / df.shape[0])

print(df[(df['test preparation'] == 'completed') & 
         (df['math score'] >= passmark)].shape[0] 
      / df[df['test preparation'] == 'completed'].shape[0])

print(df[(df['gender'] == 'female')
         & (df['test preparation'] == 'none') 
         & (df['math score'] < passmark)].shape[0] 
      / df[(df['gender'] == 'female')
         & (df['test preparation'] == 'none')].shape[0])

0.865
0.9217877094972067
0.20958083832335328


***Ответы:***

Доля студентов, сдавших экзамен по математике = **0.865**

Доля студентов, проходивших курс подготовки к экзамену, которые сдали экзамен по математике = **0.92**

Доля женщин, не проходивших курс подготовки к экзамену, которые не сдали экзамен по математике = **0.21**

**Задание 11. С помощью groupby выполните задания ниже. Также выведите время выполнения каждого из заданий.**
    * Для каждой этнической группы выведите средний балл за экзамен по чтению
    * Для каждого уровня образования выведите минимальный балл за экзамен по письму

In [271]:
display(df[['race/ethnicity', 'reading score']].groupby(by='race/ethnicity').mean())
display(df[['education', 'writing score']].groupby(by='education').min())

reading score
race/ethnicity               
group A             64.674157
group B             67.352632
group C             69.103448
group D             70.030534
group E             73.028571

writing score
education                        
associate's degree             35
bachelor's degree              38
high school                    15
master's degree                46
some college                   19
some high school               10

In [187]:
#измерим время выполнения команд
%timeit df[['race/ethnicity', 'reading score']].groupby(by='race/ethnicity').mean()
%timeit df[['education', 'writing score']].groupby(by='education').min()

2.08 ms ± 212 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
1.97 ms ± 26.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


**Задание 12. Выполните задание 11 с помощью циклов. Сравните время выполнения.**

In [206]:
def mn_1():
    o = []
    for i in df['race/ethnicity'].unique():
        sum = 0
        k = 0
        for j in range(0, df.shape[0]):
            if df['race/ethnicity'].iloc[j] == i:
                sum += df['reading score'].iloc[j]
                k += 1
        o.append((i, sum/k))
    return o

def mn_2():
    o =[]
    for i in df['education'].unique():
        mn = 100
        for j in range(0, df.shape[0]):
            if df['education'].iloc[j] == i:
                if mn > df['writing score'].iloc[j]:
                    mn = df['writing score'].iloc[j]
        o.append((i, mn))
    return o

print(mn_1())
print(mn_2())

%timeit mn_1()
%timeit mn_2()

[('group B', 67.35263157894737), ('group C', 69.10344827586206), ('group A', 64.67415730337079), ('group D', 70.03053435114504), ('group E', 73.02857142857142)]
[("bachelor's degree", 38), ('some college', 19), ("master's degree", 46), ("associate's degree", 35), ('high school', 15), ('some high school', 10)]
112 ms ± 12 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
133 ms ± 13.8 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [214]:
#выведем результаты времени работы в виде таблицы
inf = pd.DataFrame({"task": [1, 2],
                    "with 'groupby'": ["2.08 ms", "1.97 ms"],
                   "with 'for'": ["112 ms", "133 ms"]}).set_index('task')
inf

with 'groupby' with 'for'
task                          
1           2.08 ms     112 ms
2           1.97 ms     133 ms

Получили, что время выполнения задач с помощью цикла в несколько десятков раз больше, чем при использовании groupby, то есть метод groupby значительно быстрее. 

**Задание 13. Выведите на экран средние баллы студентов по каждому предмету в зависимости от пола и уровня образования. То есть должно получиться количество групп, равных 2 * (число уровней образования), и для каждой такой группы выыведите средний балл по каждому из предметов.**

Это можно сделать с помощью сводных таблиц (pivot_table):

https://www.kaggle.com/kamilpolak/tutorial-how-to-use-pivot-table-in-pandas

In [272]:
pd.pivot_table(df, values=['math score', 'writing score', 'reading score'], 
                                       index=['gender', 'education'], aggfunc="mean")

math score  reading score  writing score
gender education                                                   
female associate's degree   65.250000      74.120690      74.000000
       bachelor's degree    68.349206      77.285714      78.380952
       high school          59.351064      68.202128      66.691489
       master's degree      66.500000      76.805556      77.638889
       some college         65.406780      73.550847      74.050847
       some high school     59.296703      69.109890      68.285714
male   associate's degree   70.764151      67.433962      65.405660
       bachelor's degree    70.581818      68.090909      67.654545
       high school          64.705882      61.480392      58.539216
       master's degree      74.826087      73.130435      72.608696
       some college         69.009259      64.990741      63.148148
       some high school     67.840909      64.693182      61.375000

#### Задание 14. Сколько студентов успешно сдали экзамен по математике?

Создайте новый столбец в таблице df под названием Math_PassStatus и запишите в него F, если студент не сдал экзамен по математике (балл за экзамен < passmark), и P иначе.

Посчитайте количество студентов, сдавших и не сдавших экзамен по математике.

Сделайте аналогичные шаги для экзаменов по чтению и письму.

In [273]:
df['Math_PassStatus'] = np.where(df['math score'] < passmark, 'F', 'P')
df['Reading_PassStatus'] = np.where(df['reading score'] < passmark, 'F', 'P')
df['Writing_PassStatus'] = np.where(df['writing score'] < passmark, 'F', 'P')
print(df['Math_PassStatus'].value_counts())
print(df['Reading_PassStatus'].value_counts())
print(df['Writing_PassStatus'].value_counts())
df.head(10)

Math_PassStatus
P    865
F    135
Name: count, dtype: int64
Reading_PassStatus
P    910
F     90
Name: count, dtype: int64
Writing_PassStatus
P    886
F    114
Name: count, dtype: int64


gender race/ethnicity           education         lunch test preparation  \
0  female        group B   bachelor's degree      standard             none   
1  female        group C        some college      standard        completed   
2  female        group B     master's degree      standard             none   
3    male        group A  associate's degree  free/reduced             none   
4    male        group C        some college      standard             none   
5  female        group B  associate's degree      standard             none   
6  female        group B        some college      standard        completed   
7    male        group B        some college  free/reduced             none   
8    male        group D         high school  free/reduced        completed   
9  female        group B         high school  free/reduced             none   

    group  math score  reading score  writing score Math_PassStatus  \
0  group1          72             72             74               P   
1  group1          69             90             88               P   
2  group1          90             95             93               P   
3  group1          47             57             44               F   
4  group1          76             78             75               P   
5  group1          71             83             78               P   
6  group1          88             95             92               P   
7  group1          40             43             39               F   
8  group1          64             64             67               P   
9  group1          38             60             50               F   

  Reading_PassStatus Writing_PassStatus  
0                  P                  P  
1                  P                  P  
2                  P                  P  
3                  P                  F  
4                  P                  P  
5                  P                  P  
6                  P                  P  
7                  F                  F  
8                  P                  P  
9                  P                  P

865 студентов успешно сдали экзамен по математике 

#### Задание 15. Сколько студентов успешно сдали все экзамены?

Создайте столбец OverAll_PassStatus и запишите в него для каждого студента 'F', если студент не сдал хотя бы один из трех экзаменов, а иначе 'P'.

Посчитайте количество студентов, которые сдали все экзамены.

In [274]:
df['OverAll_PassStatus'] = np.where(
    (df['Math_PassStatus'] == 'P') & (df['Reading_PassStatus'] == 'P') & (df['Writing_PassStatus'] == 'P'), 'P', 'F')
print(df['OverAll_PassStatus'].value_counts()) #посчитаем кол-во студентов, которые сдали все экзамены
df.head(10)

OverAll_PassStatus
P    812
F    188
Name: count, dtype: int64


gender race/ethnicity           education         lunch test preparation  \
0  female        group B   bachelor's degree      standard             none   
1  female        group C        some college      standard        completed   
2  female        group B     master's degree      standard             none   
3    male        group A  associate's degree  free/reduced             none   
4    male        group C        some college      standard             none   
5  female        group B  associate's degree      standard             none   
6  female        group B        some college      standard        completed   
7    male        group B        some college  free/reduced             none   
8    male        group D         high school  free/reduced        completed   
9  female        group B         high school  free/reduced             none   

    group  math score  reading score  writing score Math_PassStatus  \
0  group1          72             72             74               P   
1  group1          69             90             88               P   
2  group1          90             95             93               P   
3  group1          47             57             44               F   
4  group1          76             78             75               P   
5  group1          71             83             78               P   
6  group1          88             95             92               P   
7  group1          40             43             39               F   
8  group1          64             64             67               P   
9  group1          38             60             50               F   

  Reading_PassStatus Writing_PassStatus OverAll_PassStatus  
0                  P                  P                  P  
1                  P                  P                  P  
2                  P                  P                  P  
3                  P                  F                  F  
4                  P                  P                  P  
5                  P                  P                  P  
6                  P                  P                  P  
7                  F                  F                  F  
8                  P                  P                  P  
9                  P                  P                  F

Кол-во студентов, которые сдали все экзамены равно 812

#### Задание 16. Переведем баллы в оценки

### Система перевода баллов в оценки
####    больше 90 = A
####      80-90 = B
####      70-80 = C
####      60-70 = D
####      50-60 = E
####    меньше 50 = F (Fail)

Создайте вспомогательную функцию, которая будет по среднему баллу за три экзамена выставлять оценку студенту по данным выше критериям.

Создайте столбец Grade и запишите в него оценку каждого студента.

Выведите количество студентов, получивших каждую из оценок.

**В случае, если средний балл попадает на границу между оценками (т.е. равен ровно 60, 70 или 80 баллов), вы можете интерпретировать условие на своё усмотрение (т.е. можете поставить за 60 баллов оценку D, а можете - E).**

In [275]:
def GetGrade(av_mark):
    if av_mark >= 90:
        mark = 'A'
    elif av_mark >=80:
        mark = 'B'
    elif av_mark >=70:
        mark = 'C'
    elif av_mark >=60:
        mark = 'D'
    elif av_mark >=50:
        mark = 'E'
    else:
        mark = 'F'
    return mark

#создадим столбец со средним баллом за три экзамена
df['Average mark'] = df[['math score', 'reading score', 'writing score']].mean(axis=1)
#создадим столбец с оценками применив функцию GetGrade для перевода баллов в оценку 
df['Grade'] = df['Average mark'].apply(GetGrade)
display(df.head(10))

#выведем кол-во студентов, получивших каждую из оценок
df['Grade'].value_counts().sort_index()

gender race/ethnicity           education         lunch test preparation  \
0  female        group B   bachelor's degree      standard             none   
1  female        group C        some college      standard        completed   
2  female        group B     master's degree      standard             none   
3    male        group A  associate's degree  free/reduced             none   
4    male        group C        some college      standard             none   
5  female        group B  associate's degree      standard             none   
6  female        group B        some college      standard        completed   
7    male        group B        some college  free/reduced             none   
8    male        group D         high school  free/reduced        completed   
9  female        group B         high school  free/reduced             none   

    group  math score  reading score  writing score Math_PassStatus  \
0  group1          72             72             74               P   
1  group1          69             90             88               P   
2  group1          90             95             93               P   
3  group1          47             57             44               F   
4  group1          76             78             75               P   
5  group1          71             83             78               P   
6  group1          88             95             92               P   
7  group1          40             43             39               F   
8  group1          64             64             67               P   
9  group1          38             60             50               F   

  Reading_PassStatus Writing_PassStatus OverAll_PassStatus  Average mark Grade  
0                  P                  P                  P     72.666667     C  
1                  P                  P                  P     82.333333     B  
2                  P                  P                  P     92.666667     A  
3                  P                  F                  F     49.333333     F  
4                  P                  P                  P     76.333333     C  
5                  P                  P                  P     77.333333     C  
6                  P                  P                  P     91.666667     A  
7                  F                  F                  F     40.666667     F  
8                  P                  P                  P     65.000000     D  
9                  P                  P                  F     49.333333     F

Grade
A     52
B    146
C    261
D    256
E    182
F    103
Name: count, dtype: int64